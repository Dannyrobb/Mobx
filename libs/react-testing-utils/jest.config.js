module.exports = {
  displayName: 'react-testing-utils',
  preset: '../../jest.preset.js',
  transform: {
    '^.+\\.[tj]sx?$': 'babel-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/libs/react-testing-utils',
};
