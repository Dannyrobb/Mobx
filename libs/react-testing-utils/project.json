{
  "root": "libs/react-testing-utils",
  "sourceRoot": "libs/react-testing-utils/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "build": {
      "executor": "@nrwl/web:rollup",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/libs/react-testing-utils",
        "tsConfig": "libs/react-testing-utils/tsconfig.lib.json",
        "project": "libs/react-testing-utils/package.json",
        "entryFile": "libs/react-testing-utils/src/index.ts",
        "external": ["react/jsx-runtime"],
        "rollupConfig": "@nrwl/react/plugins/bundle-rollup",
        "compiler": "babel",
        "assets": [
          {
            "glob": "libs/react-testing-utils/README.md",
            "input": ".",
            "output": "."
          }
        ]
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/react-testing-utils/**/*.{ts,tsx,js,jsx}"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/libs/react-testing-utils"],
      "options": {
        "jestConfig": "libs/react-testing-utils/jest.config.js",
        "passWithNoTests": true
      }
    }
  }
}
