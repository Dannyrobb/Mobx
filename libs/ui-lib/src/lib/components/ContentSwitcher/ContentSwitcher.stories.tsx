import { Story, Meta } from '@storybook/react';

import { ContentSwitcher, ContentSwitcherProps } from './ContentSwitcher';

export default {
  title: 'UI Lib / Content Switcher',
  component: ContentSwitcher,
  argTypes: {},
} as Meta;

const Template: Story<ContentSwitcherProps> = (args) => (
  <div style={{ margin: 30 }}>
    <ContentSwitcher {...args} />
  </div>
);

export const base = Template.bind({});
base.args = { options: [] };

export const prefilledOptions = Template.bind({});
prefilledOptions.args = { options: ['Asdsadsa', 'Dddsdsdsdsdsd', 'Yoyoooyoyyoyo'] };

export const prefilledOptionsDisabled = Template.bind({});
prefilledOptionsDisabled.args = { options: ['asdsadsa', 'dddsdsdsdsdsd', 'yoyoooyoyyoyo'], disabled: true };
