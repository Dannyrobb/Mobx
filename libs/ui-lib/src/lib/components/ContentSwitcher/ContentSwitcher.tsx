import * as React from 'react';

import { theme } from '@cellxpert/theme';
import styled from '@emotion/styled';

export interface ContentSwitcherProps {
  options: string[];
  selected?: string;
  disabled?: boolean;
  /**
   * @default medium
   */
  size?: 'small' | 'medium' | 'large';
}

const TabGroup = styled('div')<Pick<ContentSwitcherProps, 'options'>>(({ options }) => ({
  display: 'grid',
  gridTemplateColumns: `repeat(${options.length}, 1fr)`,
}));

const Tab = styled('button')<
  {
    size: ContentSwitcherProps['size'];
    selected?: boolean;
    disabled?: ContentSwitcherProps['disabled'];
    index: number;
  } & Pick<ContentSwitcherProps, 'options'>
>(({ selected, size = 'medium', disabled, index, options }) => ({
  ...theme.typography.meta.body2,
  display: 'flex',
  justifyContent: 'space-between',
  padding: `${theme.gutters.base * 1.5}px 0px`,
  alignItems: 'center',
  backgroundColor: selected ? theme.palette.text.main : theme.palette.additional.grey[200],
  color: selected ? theme.palette.text.contrast : theme.palette.text.main,
  border: 'none',
  '&:hover': {
    cursor: 'pointer',
    background: selected ? theme.palette.text.main : theme.palette.additional.grey[300],
  },
  '&:focus': {
    boxShadow: selected
      ? `0px 0px 0px 1px ${theme.palette.text.contrast}, 0px 0px 0px 3px ${theme.palette.main.primary1.base}`
      : `0px 0px 0px 2px ${theme.palette.main.primary1.base}`,
    zIndex: 1,
    outline: 'none',
  },
  ':first-child': {
    borderRadius: `25px 0px 0px 25px`,
  },
  ':last-child': {
    borderRadius: `0px 25px 25px 0px`,
  },
  ...(size === 'small' && {
    padding: `${theme.gutters.base * 1}px 0px`,
  }),
  ...(size === 'large' && {
    padding: `${theme.gutters.base * 2}px 0px`,
  }),
  ...(disabled && {
    pointerEvents: 'none',
    color: theme.palette.additional.grey[400],
  }),
  '::after': {
    content: index < options.length - 1 ? '""' : 'none',
    background: selected ? theme.palette.text.main : theme.palette.additional.grey[300],
    left: '0px',
    height: size === 'small' ? '16px' : size === 'medium' ? '20px' : '24px',
    width: '1px',
    boxSizing: 'border-box',
  },
}));

const TabLabel = styled('span')(() => ({
  padding: `0px ${theme.gutters.base * 2}px`,
}));

export const ContentSwitcher: React.FC<ContentSwitcherProps> = ({ options, selected, size = 'medium', disabled }) => {
  const [selectedContent, setSelectedContent] = React.useState(selected ? selected : options[0]);
  console.log(selectedContent);

  return (
    <>
      <TabGroup {...{ options }}>
        {options.map((item, index) => (
          <Tab
            key={item}
            {...{
              options,
              index,
              size,
              selected: selectedContent === item,
              onClick: () => setSelectedContent(item),
              disabled,
            }}
          >
            <TabLabel>{item}</TabLabel>
          </Tab>
        ))}
      </TabGroup>
    </>
  );
};

export default ContentSwitcher;
