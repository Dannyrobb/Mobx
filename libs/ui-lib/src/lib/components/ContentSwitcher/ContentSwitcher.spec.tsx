import { render, fireEvent } from '@cellxpert/react-testing-utils';

import { ContentSwitcher } from './ContentSwitcher';

describe('ContentSwitcher', () => {
  it('should render successfully', () => {
    const { baseElement } = render(<ContentSwitcher options={['left', 'middle', 'right']} />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should show the first item as selected by default', () => {
    const { getByText } = render(<ContentSwitcher options={['left', 'middle', 'right']} />);

    // TODO - show debug, add the test to click on the second item, try moving the test-setup to the testind utils

    const firstElement = getByText('left').closest('button');
    expect(firstElement).toMatchSnapshot();
    console.log({ firstElement: firstElement?.style });

    expect(firstElement).toHaveStyle('background-color: rgb(29, 29, 27)');
  });

  it('should be able to change tabs', () => {
    const { getByText } = render(<ContentSwitcher options={['left', 'middle', 'right']} />);

    const secondElement = getByText('middle').closest('button');
    expect(secondElement).toMatchSnapshot();

    expect(secondElement).not.toHaveStyle('background-color: rgb(29, 29, 27)');
    // TODO = add assertion for actual background color

    fireEvent.click(secondElement!);
    expect(secondElement).toHaveStyle('background-color: rgb(29, 29, 27)');
  });

  it.todo('should support disabled tabs');
});
