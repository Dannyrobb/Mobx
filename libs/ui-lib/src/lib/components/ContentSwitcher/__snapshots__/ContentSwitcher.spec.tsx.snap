// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentSwitcher should be able to change tabs 1`] = `
<button
  class="css-4hccwr-Tab e1klf5xo1"
>
  <span
    class="css-1lhcv4-TabLabel e1klf5xo0"
  >
    middle
  </span>
</button>
`;

exports[`ContentSwitcher should render successfully 1`] = `
<body>
  <div>
    <div
      class="css-161nnu1-TabGroup e1klf5xo2"
    >
      <button
        class="css-dczsz7-Tab e1klf5xo1"
      >
        <span
          class="css-1lhcv4-TabLabel e1klf5xo0"
        >
          left
        </span>
      </button>
      <button
        class="css-4hccwr-Tab e1klf5xo1"
      >
        <span
          class="css-1lhcv4-TabLabel e1klf5xo0"
        >
          middle
        </span>
      </button>
      <button
        class="css-1xif5kq-Tab e1klf5xo1"
      >
        <span
          class="css-1lhcv4-TabLabel e1klf5xo0"
        >
          right
        </span>
      </button>
    </div>
    <svg
      class="css-9qkvca-SvgIconBase eb84nrk0"
      fill="#126AEC"
      height="16"
      viewBox="0 0 32 32"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
      xmlns:xlink="http://www.w3.org/1999/xlink"
    >
      <rect
        fill="none"
        height="32"
        id="verticalDivider"
        width="32"
      />
      <rect
        height="16"
        transform="rotate('90')"
        width="2"
        x="8"
        y="15"
      />
    </svg>
  </div>
</body>
`;

exports[`ContentSwitcher should show the first item as selected by default 1`] = `
<button
  class="css-dczsz7-Tab e1klf5xo1"
>
  <span
    class="css-1lhcv4-TabLabel e1klf5xo0"
  >
    left
  </span>
</button>
`;
